Allow incomplete return types in function declarations.

zygoloid opened the pull request:
Move completeness check to the point where the function is defined or first called. This means we also defer deciding whether the function has a return slot until that point. Instead of storing a return slot per function, store the location of the return storage, which may or may not be used, and compute and store a separate flag saying whether to use it at the point of first use or definition.

This is the final piece in supporting simple `Make` functions in classes as a replacement for constructors.

zygoloid committed:
Allow incomplete return types in function declarations.

Move completeness check to the point where the function is defined or
first called. This means we also defer deciding whether the function has
a return slot until that point. Instead of storing a return slot per
function, store the location of the return storage, which may or may not
be used, and compute and store a separate flag saying whether to use it
at the point of first use or definition.

jonmeow commented on toolchain/sem_ir/function.h:
```diff
@@ -52,14 +76,30 @@ struct Function : public Printable<Function> {
+  // point where the function is defined.
+
+  // Whether the function uses a return slot.
+  enum class ReturnSlot : int8_t {
```
Declaration order. I know mixed feelings on this, but I think we should be following the style guide in spots like this.

https://google.github.io/styleguide/cppguide.html#Declaration_Order

zygoloid replied:
Sure, done.

jonmeow commented on toolchain/check/import_ref.cpp:
```diff
@@ -733,8 +730,10 @@ class ImportRefResolver {
+         .return_storage_id = new_return_storage,
          .is_extern = function.is_extern,
+         // There has been no call or definition in this file.
+         .return_slot = SemIR::Function::ReturnSlot::NotComputed,
```
Should this be forwarded when Absent or similar? If it's an Error in an import, should it become valid when imported? i.e., if known from the API, I'm not confident it should be ignored.

As an example test:

```
library "a" api;
class C;
// Diagnoses incomplete type, right?
fn F() -> C;
```

```
library "b" api;
import library "a";

fn G() {
  // Will this diagnose incomplete?
  F();
}
```

```
library "c" api;
import library "a";

class C {}

fn H() {
  // Does this get a different result for the return slot?
  F();
}
```

Note, that'd also have implications for merging, since it's not part of the prev_function merge right now.

zygoloid replied:
When the return type is incomplete, the `Error` state gets re-checked on every call anyway; the difference between `NotChecked` and `Error` is that `Error` suppresses diagnostics. So I think it's just a question of whether we want a diagnostic per file, or one in total.

Per in-person discussion, more error suppression is probably better, so changed this to import and merge the return slot information.

jonmeow commented on toolchain/check/function.cpp:
```diff
@@ -327,4 +327,42 @@ auto MergeFunctionRedecl(Context& context, SemIR::LocId loc_id,
+                                   IncompleteTypeInFunctionReturnType,
+                                   function.return_type_id);
+  };
+  std::optional<llvm::function_ref<auto()->Context::DiagnosticBuilder>>
```
Would it be worth a typedef on Context for this?

zygoloid replied:
Rearranged to avoid the need for this type.

jonmeow commented on toolchain/check/function.cpp:
```diff
@@ -315,7 +315,7 @@ auto MergeFunctionRedecl(Context& context, SemIR::LocId loc_id,
     prev_function.param_refs_id = new_function.param_refs_id;
     prev_function.return_type_id = new_function.return_type_id;
-    prev_function.return_slot_id = new_function.return_slot_id;
+    prev_function.return_storage_id = new_function.return_storage_id;
```
Would it be worth asserting new_function.return_slot is NotComputed? (tieing in with my other comment that maybe it shouldn't always be NotComputed)

zygoloid replied:
Changed to merge the slot info.

jonmeow reviewed (approved):
Approving, though not sure about the behavior on import.

zygoloid committed:
Address review comments.

Move the location for the diagnostic for a call with an incomplete
return type to the call, rather than pointing the diagnostic at the
function declaration.

zygoloid committed:
Avoid awkward type.

jonmeow reviewed (approved):
Thanks, all the changes look great.

jonmeow commented:
Might be a merge conflict to investigate -- tests look like they need to be updated.

zygoloid committed:
Merge branch 'trunk' into toolchain-return-self

zygoloid committed:
Autoupdate
